### PR TITLE
Improve particle performance

### PR DESCRIPTION
**Pull request by mucaho**
* Extend API to restart particle animation with previously used options.
* Minimize particle object allocation - reuse particles from particle pool instead.
* Flatten particle vector properties into plain properties.
* Split particle color update from rendering better, to facilitate easier rendering with WebGL backend.
* Simplify and remove some redundant code.

Planned for next PR: Change particles to be a view into a TypedArray (as discussed in #1132 )

### CODE REVIEW

File: src/graphics/particles.js [hunk 1]
```diff
@@ -146,223 +160,232 @@ Crafty.c("Particles", {
...
+                    particle.colourR = c > 255 ? 255 : c < 0 ? 0 : ~~c;
+                    c = particle.colourG + particle.deltaColourG;
+                    particle.colourG = c > 255 ? 255 : c < 0 ? 0 : ~~c;
+                    c = particle.colourB + particle.deltaColourB;
+                    particle.colourB = c > 255 ? 255 : c < 0 ? 0 : ~~c;
+                    c = particle.colourA + particle.deltaColourA;
+                    particle.colourA = c > 1 ? 1 : c < 0 ? 0 : (~~(c * 100)) / 100;
+
+                    // Decrease particle's lifespan
+                    particle.timeToLive--;
```
Comment (starwed) [inline src/graphics/particles.js:311]: With the current implementation, this is really `framesToLive` -- at some point it should get updated to use the proper `dt`.  That would maybe involve changing how velocity is defined.

**Review by starwed — APPROVED** [review]
Looks good -- this should help the performance of particles quite a lot.

File: src/graphics/particles.js (hunk 1 above)
Comment (mucaho) [reply src/graphics/particles.js:311]: Yep, for further roadmap, see the [particle fx project](https://github.com/craftyjs/Crafty/projects/3)

**Comment by mucaho** [comment]
Thanks for taking a look at it!

**Pull request merged**